git-grep UI

gerhardol opened the pull request:
Fixes #4912
Depends on #11590 , only 92d937e21c88295b37be2f760ce91560c2e9ddb3 (and forward if added later) to be reviewed

## Proposed changes

An UI for git-grep, to search commits without checking out those commits. (When working with this I have often used it to view worktree too.)
https://git-scm.com/docs/git-grep

The UI in this PR adds a popup as well as a searchbox to the Diff tab.

The searchbox dropdown has the search history for the GE session, not persistent.
The popup and searchbox are kept in sync.

Search in the popup is done when button is pressed (or Enter),
the diff tab box is searching when typing (similar in the filter).
If there is no `-e ` in the search string, this is added to the search string.
Most git-grep options can be added, including  `--and` etc.

git-grep options to handle case insensitive is available, but most options can be added in the searchbox.
Persistent options can be added too for instance `--show-function`.

The output of git-grep is always for a commit, including index and worktree.
Note that the selected file in the grep group is not always followed when a new commit is selected, as it is not the primary group. This could be specially handled (but "grep" knowledge need to be added to RevDiff).

Performance
In the Git repo, the git-grep requires about 200 ms for me, the information is added async after normal diff so no visible.
For the Linux repo, git-grep can require several seconds. As the current search is aborted when a new is started, works fine for me.
The processing in the fileviewer is faster than for normal diffs.

Limitations are basically the same as #11590 ,missing
* Documentation for the changes, how colors etc can be changed by Git config
* tests for the new functionality. May be postponed to a later PR

## Screenshots <!-- Remove this section if PR does not change UI -->

### After

The popup window and the optional FileStatusList combobox
![image](https://github.com/gitextensions/gitextensions/assets/6248932/5d30eb67-2272-41ac-b5a8-d5d946311c3b)

Example usage.
![grep-example](https://github.com/gitextensions/gitextensions/assets/6248932/f458a445-4104-4cec-8968-fe798dfe2afe)

## Test methodology <!-- How did you ensure quality? -->

Manual

## Merge strategy

squash.

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

vbjay commented:
Just a point. Not saying git grep shouldn't be used. Just saying ripgrep is very speedy. So if searching working directory vs git history.

gerhardol commented:
> Just a point. Not saying git grep shouldn't be used. Just saying ripgrep is very speedy. So if searching working directory vs git history.

To clarify if someone is wondering: This feature is primarily to search Git commit history without a checkout.
 
This feature is not really competing with searching in the worktree. Searching in the editor and specific tools will normally be a better choice.
I doubt anyone want to implement a GUI with all search options even (command line options can be provided though).
I have ended up using the feature in the worktree too (for instance a quick way to search and exclude submodules and ignored files) but this is a sideeffect.

gerhardol commented:
If noone has any other preferences, I plan to add an extra searchbox to diff.

mstv commented:
My 2ct: This feature seems to be pretty related to the "Find" function - particularly to "Find all" in IDEs.
I would add it to the FindAndReplaceForm (or else have a new one opened using `Shift+Ctrl+F`). "Find & highlight all" does not work (and does not really apply) if invoked from the FileStatusList. So this button could be replaced with "Find all (git grep)" in this case.
The result as diff group feels appropriate.
There should be a way to discard the grep group.

I guess I would use it more often if the search could be limited to the changes instead of searching the entire file tree of the selected commit (which can be useful in other cases).

The detection of "-e" does not work in all locations. I have tried to ignore the character case:

![image](https://github.com/gitextensions/gitextensions/assets/36601201/719a0957-95e1-4e1b-8d9c-32878a848b14)

gerhardol commented:
> My 2ct: This feature seems to be pretty related to the "Find" function - particularly to "Find all" in IDEs. 

It is similar to Find all, but only for the files in Git and only for the selected commit, you do not have to check out the commits.

> "Find & highlight all" does not work (and does not really apply) if invoked from the FileStatusList. 

"Find & highlight all" is only for current file, should be renamed (moved to context menu in fileviewer?).

> I would add it to the FindAndReplaceForm (or else have a new one opened using `Shift+Ctrl+F`). .... So this button could be replaced with "Find all (git grep)" in this case. The result as diff group feels appropriate. There should be a way to discard the grep group.

git-grep works quite different from current find, so a separate popup is probably better than a combined form - maybe a button to switch to the other form.
A context menu item and keyboard shortcut to clear the search maybe.
`Shift+Ctrl+F` would be the logical shortcut, but it will hide the "find file".

But I may prefer the searchbox anyway, you can search interactively, get feedback faster.

> I guess I would use it more often if the search could be limited to the changes instead of searching the entire file tree of the selected commit (which can be useful in other cases).

git-grep cannot search the diff, that would be a completely different algorithm.
It could be possible to combine diff/grep, but the result would maybe be unexpected.

> The detection of "-e" does not work in all locations. I have tried to ignore the character case:

Will try to tweak this.
The options could be provided separately too. (But I do not want to implement all options in a gui, including and/or, overkill.)

mstv commented:
> But I may prefer the searchbox anyway, you can search interactively, get feedback faster.
> git-grep cannot search the diff, that would be a completely different algorithm.

I prefer code browsing in an IDE ("go to definition", "browse references", "exclude unit tests"). So I guess I will use grep rarely and rather save the space in favor of the Git Output Control.
A toolbar button could be added to the Diff search bar in order to switch the toolbar mode to grep or else to show & hide an additional grep bar.

> `Shift+Ctrl+F` would be the logical shortcut, but it will hide the "find file".

I think this is acceptable for the Diff tab - especially as grep is kind of a generalization of "find file". And one can quickly switch to the File Tree.

gerhardol commented:
Implemented a searchbox.
I believe this ready for review and inclusion, but there are some discussion points:

* mstv wanted to search with a popup instead. That could be added in addition to the searchbox, the box could be hidden.
* There are no tests. Will adapt FileStatusListTest.cs if this is accepted, but the tests depends on how it is implemented.
* If #11431 is merged, I prefer to add the methods in GiTModule to some other module (it is shared FileStatusList, GitUIExtensions and GrepHighlightService.

mstv commented on GitUI/UserControls/FileStatusList.Designer.cs:
```diff
@@ -151,18 +150,59 @@ private void InitializeComponent()
+            SearchWatermarkLabel.Location = new Point(4, 27);
+            SearchWatermarkLabel.Name = "SearchWatermarkLabel";
+            SearchWatermarkLabel.Size = new Size(184, 13);
+            SearchWatermarkLabel.Text = "Search files in revision using a regular expression...";
```
❔
```suggestion
            SearchWatermarkLabel.Text = "Search text in all files of revision using a regular expression...";
```
might be clearer

mstv reviewed (commented):
> Will squash if mstv agrees

I have not looked into the code yet.

> mstv wanted to search with a popup instead. That could be added in addition to the searchbox, the box could be hidden.

My main point is that I do not want this searchbox to consume space all the time because I will rarely use it.
I admit it is really quick and it can be (mis-)used to get all files for the artificial commits (by searching for ".").

Could we add a toggle button right of the `Filter files` box?
Though it would need to remain visible if there are "No changes".

... which does not look good (distracting) in the current state IMO:

![image](https://github.com/gitextensions/gitextensions/assets/36601201/8581a49a-6c8c-413f-9158-be0501995a29)

gerhardol commented:
> > Will squash if mstv agrees

> I have not looked into the code yet.

If you have not reviewed the previous code, I assume I can squash.

> > mstv wanted to search with a popup instead. That could be added in addition to the searchbox, the box could be hidden.
> 
> My main point is that I do not want this searchbox to consume space all the time because I will rarely use it. I admit it is really quick and it can be (mis-)used to get all files for the artificial commits (by searching for ".").
> 
> Could we add a toggle button right of the `Filter files` box? Though it would need to remain visible if there are "No changes".

I rather have an enable toggle in the context menu as for *Show file differences*
(Ctrl-shift-f could be shortcut)

popup would be a way to configure the options though, maybe additional control.

mstv commented:
Having it in one line with "::" feels too cryptic to me.
A toggle button in the upper right corner would make it discoverable. (Please use `RuntimeSetting` in order to not interfere with other running instances.)
Default search options could be set using a context menu (to be added in a follow-up).

![image](https://github.com/gitextensions/gitextensions/assets/36601201/f2baa7e1-b4a1-4a0b-a134-c002b15f5fec)

![image](https://github.com/gitextensions/gitextensions/assets/36601201/954e2acd-0b49-4304-8163-e01f0a42ce00)

In case of "no changes", I would always hide the search bar - but not change the setting.

gerhardol commented:
> Having it in one line with "::" feels too cryptic to me. A toggle button in the upper right corner would make it discoverable. (Please use `RuntimeSetting` in order to not interfere with other running instances.) Default search options could be set using a context menu (to be added in a follow-up).

I am adding a configuration, default off, maybe even a popup.
Depends on #11472 though, setting this to draft again.

I want the options to be persistent though, not just runtime.
The filter items are instance local, could be be runtime settings.

![image](https://github.com/gitextensions/gitextensions/assets/6248932/376fe57b-4f97-4017-8a3d-071615ffaf98)

> In case of "no changes", I would always hide the search bar - but not change the setting.

Yes, in the same way as the filterbox is hidden if no changes.

mstv commented:
> I want the options to be persistent though, not just runtime. The filter items are instance local, could be be runtime settings.

That's what `RuntimeSetting<>` is intended for:
- no side-effect on other instances
- default persisted by means of

![image](https://github.com/gitextensions/gitextensions/assets/36601201/3e7670ed-3c38-43b4-89ba-609dfdae2066)

---
 
> ![image](https://private-user-images.githubusercontent.com/6248932/293357747-376fe57b-4f97-4017-8a3d-071615ffaf98.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDM4NTI2OTksIm5iZiI6MTcwMzg1MjM5OSwicGF0aCI6Ii82MjQ4OTMyLzI5MzM1Nzc0Ny0zNzZmZTU3Yi00Zjk3LTQwMTctOGEzZC0wNzE2MTVmZmFmOTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDIzMTIyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMjlUMTIxOTU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjFlM2I3YjQ3ZTk5MWMyOThhNGFlZDM4MTE5MWU5ZWQ4ZDhkMTQ4ZDcxMTI5ODdjOTBkMzFhZjVjOWIyZjNjOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.0UZvCLd1Dfb2vACFsWEgePYUhnyJcqofCG2NnePNEfg)

- discoverable with default value `true`
- insert below of "Find"?
- will need a description how to hide it by default

> > In case of "no changes", I would always hide the search bar - but not change the setting.
> 
> Yes, in the same way as the filterbox is hidden if no changes.

Not exactly, the search box still makes sense for empty commits. The user should be able to activate it.

mstv commented:
If the "search commit" combobox is visible, the "filter files" combobox should always be shown, too.
"No changes" shall be hidden, if there are search results.

![grafik](https://github.com/gitextensions/gitextensions/assets/36601201/628c65d5-a42c-467d-b9f2-032519aec48f)

mdonatas commented:
I've incorporated this branch into my custom build and have already made use of it at least 3 times. Loving it! At the same time I am getting strong vibes that GitExtensions is turning in the kind of software where one needs to watch a youtube video or someone else using it to learn of these power-user features. Regardless, I still welcome this feature!

gerhardol commented:
Pushed an update. This PR is now dependent on #11590, it uses the Git coloring. But I believe this PR works as I intend it too.
I do not plan to change this PR much more but a follow up should add tests and documentation. A popup to search in addition to the searchbox in revdiff should be added too, also to set options.

gerhardol commented:
Implemented a popup to search in (with the searchbox optional). Pushed to tmp/git-coloring. If #11590 is progressing, I will update this PR, including screenshots.

mdonatas commented:
Also the settings page
![image](https://github.com/gitextensions/gitextensions/assets/483659/a5dd7bf7-e9fe-47b3-998e-222bbca3d248)

gerhardol commented:
> The grey placeholder text label problem (oversized) was there before too, but that didn't manifest visibly until this change.

The label padding must be changed from 4->2 (but 15+2*4=23 so why is padding counted twice?) (OK with 100% too)

For the extra space I do not know, toggling the box mostly fixes the appearance (except for a smaller margin to the right).

> Also the settings page

Fixing

mdonatas commented:
I've manged to fix the label in my build by reverting these lines from a pretty recent PR https://github.com/gitextensions/gitextensions/pull/11472/files#diff-5cb23b1d441e66ddc2a3b26759cf7c0119dcf370010baff9cb2654611bdd891aR121-R124

gerhardol commented:
> I've manged to fix the label in my build by reverting these lines from a pretty recent PR https://github.com/gitextensions/gitextensions/pull/11472/files#diff-5cb23b1d441e66ddc2a3b26759cf7c0119dcf370010baff9cb2654611bdd891aR121-R124

That had other issues I believe.
Fix (as well for settings) in tmp/git-coloring for now as I await that PR to be approved before updating this dependent PR again

mdonatas commented:
Just wanted to add my 0.02€ that green color for highlighting looked much better (seen it when using older versions of this feature) as it gave out positive vibes of things found. Red color on the contrary gives _errory_ vibes that things were removed or are missing.

![image](https://github.com/gitextensions/gitextensions/assets/483659/171a3ceb-028d-480e-9f31-0a01d02d5b9c)

gerhardol commented:
Pushed a hack for the gap between the search and filter boxes.
There is a minor display issue still if the search box is shown after startup, fixed with a toggle.

Another fix is to not persist the searchbox visibility after startup...
![image](https://github.com/gitextensions/gitextensions/assets/6248932/26cd5362-5b06-4eb6-ae21-3318794dc2b2)
![image](https://github.com/gitextensions/gitextensions/assets/6248932/f8f432c9-43a6-48f9-b121-7ebe0cfd0bb0)

The update rebases the changes for #11590 too (that PR itself is not rebased).

mdonatas commented:
Are you also getting this error upon startup in `gerhardol/feature/i4912-git-grep-ui`?
I've even did a `git clean -xdf` to make sure that it's not my local artifacts causing this.

![image](https://github.com/gitextensions/gitextensions/assets/483659/2955e5d7-6f60-4577-8ae2-206c319fcdec)

gerhardol commented:
> Are you also getting this error upon startup in `gerhardol/feature/i4912-git-grep-ui`? I've even did a `git clean -xdf` to make sure that it's not my local artifacts causing this.

I have seen it occasionally, not right now. (A recent PR should have handled that.)

gerhardol commented:
> Are you also getting this error upon startup in `gerhardol/feature/i4912-git-grep-ui`? I've even did a `git clean -xdf` to make sure that it's not my local artifacts causing this.

A similar issue was fixed in #11533 @mstv any clues?

gerhardol commented:
An alternative to 6050919c63b0c3bc220fe60fd516656305fa7c7a would be to not toggle search off/on while loading the control, as the size is apparently incorrect at that point.
This would require that 8 occurrences that will never use the searchbox explicitly disables it or that the constructor gets an optional argument (the designer have problems with that?)

mstv commented:
It is similar to #11533 but not that issue which has been addressed by it.
The highlighted error message _gives_ the clue. (I have not seen it myself.)
The full callstack will make it clearer.

mdonatas commented:
I think I've figured out the error mentioned previously https://github.com/gitextensions/gitextensions/pull/11350#issuecomment-1972881352

It's only happening on a non-100% scaling factor. I get it every single time.

First-order problem is that `Load` event is triggered for some deeply nested controls while parent controls are not yet initialized. And it seems this is happening only if scaling is needed.
![image](https://github.com/gitextensions/gitextensions/assets/483659/c1206644-7c9b-426d-be4b-f1c65da3ae67)

I've then narrowed it down to a `SearchWatermarkLabel` label in the `FileStatusList` control. A problem **I think** is that it's made visible too early by
`FileStatusList.SearchEnabledForList > EnableSearchForList > SetFileStatusListVisibility > SetSearchWatermarkLabelVisibility`

and that is because `RevisionDiffControl` makes it visible a little bit too early in the process
![image](https://github.com/gitextensions/gitextensions/assets/483659/74aad0dc-3436-47c4-9f47-b7415c6332a8)

If we moved `DiffFiles.SearchEnabledForList = true` outside of the `InitializeComponent` and set it right after it returns in the constructor of the `RevisionDiffControl` then there is no error
![image](https://github.com/gitextensions/gitextensions/assets/483659/1981ba42-8ca5-4129-80f2-cbed80ede21c)

I don't know if it ought to work even with how it's done currently but this is a fix/work-around I've found.

mdonatas commented:
Whoops! This ☝️ "fix" break watermark label scaling/sizes :/

mdonatas commented:
Alright.. so I've went through the callstack looking at the path leading to `Load` being called.. looking for how to break this flow. One easy alas whacky way is...

first of all look at this WinForms code on the way to `Load` being called. At line `2661` this GETTER!!! leads down the path of creating controls prematurely etc... also notice an `if` above it which check for mnemonics being enabled..
![image](https://github.com/gitextensions/gitextensions/assets/483659/b1d604a9-8e43-4f69-9426-e1f733e321bd)

this is stupid, but this fixes the error as we prevent the logic flow from reaching the `!control.ShowKeyboardCues` part 🤦
![image](https://github.com/gitextensions/gitextensions/assets/483659/8274fd99-b54b-4d11-be71-fdf0cbaf16a7)

mdonatas commented:
After the latest commit high-dpi looks 👌 and the error on `Load` is gone 👍

gerhardol commented:
> I don't know if it ought to work even with how it's done currently but this is a fix/work-around I've found.

> Whoops! This ☝️ "fix" break watermark label scaling/sizes :/

> After the latest commit high-dpi looks 👌 and the error on Load is gone 👍

If the workaround I added for the resizing is removed, this seem to work!
Still the width for the filter box has to be set.

Also fixed that SearchCommit popup is not on top of Commit (same as for Find)

gerhardol commented:
Squashed and rebased on master

mstv commented on GitUI/Editor/Diff/GrepHighlightService.cs:
```diff
@@ -0,0 +1,222 @@
+                    DebugHelpers.Fail($"Cannot parse lineNo for grep {line} ({sb.Length})");
+                }
+
+                // git-grep emitts an enpty line last, should not be displayed.
```
```suggestion
                // git-grep emits an empty line last, should not be displayed.
```

mstv commented on GitUI/Editor/Diff/GrepHighlightService.cs:
```diff
@@ -0,0 +1,222 @@
+                    Color color = AppColor.DiffAddedExtra.GetThemeColor();
+                    Color forecolor = ColorHelper.GetForeColorForBackColor(color);
+                    if (AnsiEscapeUtilities.TryGetTextMarker(new()
+                    {
+                        DocOffset = sb.Length + column - 1,
+                        Length = grepLength,
+                        BackColor = color,
+                        ForeColor = forecolor
+                    },
```
```suggestion
                        {
                            DocOffset = sb.Length + column - 1,
                            Length = grepLength,
                            BackColor = color,
                            ForeColor = forecolor
                        },
```

mstv commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -374,6 +374,16 @@ public ArgumentString GetExtraDiffArguments(bool isRangeDiff = false)
+            return new ArgumentBuilder
+            {
+                "-h",
+                { ShowEntireFile, "--context=9000", $"--context={NumberOfContextLines}" },
```
special magic number, let's use a more insane value
```
int numberOfContextLines = ShowEntireFile ? 100_000 : NumberOfContextLines;
```

gerhardol replied:
OK, this is not changed from before

mstv commented on GitUI/Editor/FileViewerInternal.cs:
```diff
@@ -387,6 +388,28 @@ public void AddTextHighlighting()
+        /// Get the next/previous line for the grep match.
+        /// </summary>
+        /// <param name="next"><c>true</c> if next position, <c>false</c> if previous.</param>
+        /// <returns><c>true</c> if.</returns>
+        public bool GetMatchLine(bool next)
+        {
+            if (_textHighlightService is not GrepHighlightService grepHighlightService)
+            {
+                return false;
+            }
+
+            if (!(grepHighlightService.GetGrepLineNum(LineAtCaret, next) is int line) || line < 0)
+            {
+                return true;
```
This will answer whether the second `return` shall really be `true`.

gerhardol replied:
Search is simplified in #11628, special case for grep is removed

gerhardol replied:
I could remove search in this PR to reduce noise if you prefer.

mstv replied:
I should have started with "xmldoc is incomplete". This was the main point.

gerhardol replied:
Navigate is still included

mstv commented on GitUI/GitUIExtensions.cs:
```diff
@@ -98,7 +98,29 @@ public static partial class GitUIExtensions
+                }
 
+                await fileViewer.ViewGrepAsync(item, text: result.StandardOutput, useGitColoring: AppSettings.UseGitColoring.Value, item.Item.GrepString);
                 return;
```
```suggestion
``

mstv commented on GitUI/UserControls/FileStatusDiffCalculator.cs:
```diff
@@ -277,13 +299,34 @@ DiffBranchStatus GetDiffStatus(GitItemStatus f, bool atoBDiff)
+                return null;
+            }
+
+            IReadOnlyList<GitItemStatus>? statuses = GetModule().GetGrepFilesStatus(selectedRev, _fileStatusDiffCalculatorInfo.GrepArguments, cancellationToken);
+            if (statuses is null)
+            {
+                // error executing Git, ignore
+                return null;
+            }
+
```
```suggestion
            IReadOnlyList<GitItemStatus> statuses = GetModule().GetGrepFilesStatus(selectedRev, _fileStatusDiffCalculatorInfo.GrepArguments, cancellationToken);
```

mstv commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -728,7 +782,7 @@ public void SetDiffs(IReadOnlyList<GitRevision> revisions)
             _enableDisablingShowDiffForAllParents = true;
             _diffCalculator.SetDiff(revisions, headId: null, allowMultiDiff: false);
-            UpdateFileStatusListView(_diffCalculator.Calculate(cancellationToken));
+            UpdateFileStatusListView(_diffCalculator.Calculate(Array.Empty<FileStatusWithDescription>(), refreshDiff: true, refreshGrep: false, cancellationToken));
```
```suggestion
            UpdateFileStatusListView(_diffCalculator.Calculate(prevList: [], refreshDiff: true, refreshGrep: false, cancellationToken));
```

mstv commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -740,7 +794,21 @@ public async Task SetDiffsAsync(IReadOnlyList<GitRevision> revisions, ObjectId?
             cancellationToken.ThrowIfCancellationRequested();
             _diffCalculator.SetDiff(revisions, headId, allowMultiDiff: true);
-            IReadOnlyList<FileStatusWithDescription> gitItemStatusesWithDescription = _diffCalculator.Calculate(cancellationToken);
+            IReadOnlyList<FileStatusWithDescription> gitItemStatusesWithDescription = _diffCalculator.Calculate(Array.Empty<FileStatusWithDescription>(), refreshDiff: true, refreshGrep: false, cancellationToken);
```
```suggestion
            IReadOnlyList<FileStatusWithDescription> gitItemStatusesWithDescription = _diffCalculator.Calculate(prevList: [], refreshDiff: true, refreshGrep: false, cancellationToken);
```

mstv commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -740,7 +794,21 @@ public async Task SetDiffsAsync(IReadOnlyList<GitRevision> revisions, ObjectId?
+
+            await TaskScheduler.Default;
+            cancellationToken.ThrowIfCancellationRequested();
+            gitItemStatusesWithDescription = _diffCalculator.Calculate(GitItemStatusesWithDescription, refreshDiff: false, refreshGrep: true, cancellationToken);
```
```suggestion
            gitItemStatusesWithDescription = _diffCalculator.Calculate(prevList: GitItemStatusesWithDescription, refreshDiff: false, refreshGrep: true, cancellationToken);
```

mstv commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -1410,7 +1514,7 @@ private void FileStatusListView_ContextMenu_Opening(object sender, CancelEventAr
                     ThreadHelper.FileAndForget(async () =>
                     {
-                        IReadOnlyList<FileStatusWithDescription> gitItemStatusesWithDescription = _diffCalculator.Calculate(cancellationToken);
+                        IReadOnlyList<FileStatusWithDescription> gitItemStatusesWithDescription = _diffCalculator.Calculate(GitItemStatusesWithDescription, refreshDiff: true, refreshGrep: false, cancellationToken);
```
```suggestion
                        IReadOnlyList<FileStatusWithDescription> gitItemStatusesWithDescription = _diffCalculator.Calculate(prevList: GitItemStatusesWithDescription, refreshDiff: true, refreshGrep: false, cancellationToken);
```

mstv commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -1431,6 +1535,32 @@ private void FileStatusListView_ContextMenu_Opening(object sender, CancelEventAr
+                return;
+            }
+
+            SearchCommitForm search = new()
```
```suggestion
            using SearchCommitForm search = new()
```

gerhardol replied:
The form lives after this section. Will block multiple opening though

mstv commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -1835,6 +1965,98 @@ private void FilterComboBox_SizeChanged(object sender, EventArgs e)
+                // delay to handle keypresses
+                await Task.Delay(delay, cancellationToken);
+                _diffCalculator.SetGrep(search);
+                IReadOnlyList<FileStatusWithDescription> gitItemStatusesWithDescription = _diffCalculator.Calculate(GitItemStatusesWithDescription, refreshDiff: false, refreshGrep: true, cancellationToken);
```
```suggestion
                IReadOnlyList<FileStatusWithDescription> gitItemStatusesWithDescription = _diffCalculator.Calculate(prevList: GitItemStatusesWithDescription, refreshDiff: false, refreshGrep: true, cancellationToken);
```

mstv commented on UnitTests/GitCommands.Tests/Settings/AppSettingsTests.cs:
```diff
@@ -291,6 +291,9 @@ static IEnumerable<(PropertyInfo property, object defaultValue, bool isNullable,
                 yield return (properties[nameof(AppSettings.DiffVerticalRulerPosition)], 0, false, false);
+                yield return (properties[nameof(AppSettings.GitGrepUserArguments)], "", false, false);
+                yield return (properties[nameof(AppSettings.GitGrepIgnoreCase)], false, false, false);
+                yield return (properties[nameof(AppSettings.GitGrepMatchWholeWord)], false, false, false);
```
+`ShowSearchCommit`

gerhardol replied:
ShowAvailableDiffTools was missing too

mstv commented on Plugins/GitUIPluginInterfaces/IGitModule.cs:
```diff
@@ -445,5 +445,8 @@ public interface IGitModule
 
         GitBlame Blame(string? fileName, string from, Encoding encoding, string? lines = null, CancellationToken cancellationToken = default);
+
+        IReadOnlyList<GitItemStatus>? GetGrepFilesStatus(GitRevision revision, string greps, CancellationToken cancellationToken = default);
```
(All important Forms call this from a background task.)

gerhardol replied:
what do you mean?

mstv replied:
>> (All important Forms call this from a background task.)

> what do you mean?

This function does not need to be made `async`, (which would make its implementation more complicate.)

(In the "Files changed" tab, this function is located right above an added `async` function. The comment loses this context.)

gerhardol commented:
Thanks for the feedback

> The history contains "-e". It should also remember the other options including case and whole word.

case/whole word is now handled as the persistent options. Do you want all options here?

It may look better if the "default" search without -e is reflected in the history too.
I want to enable full flexibility but maybe not display if not needed.

mstv commented:
>> The history contains "-e". It should also remember the other options including case and whole word.

> case/whole word is now handled as the persistent options Do you want all options here?
> 
> It may look better if the "default" search without -e is reflected in the history too.
> I want to enable full flexibility but maybe not display if not needed.

I have seen the AppSettings - appreciated.

The state of these options from the previous search uses to not match the current need often - at least in _my_ experience. 😉
When switching between searches, it would be useful to restore these options, too.

gerhardol commented:
> The state of these options from the previous search uses to not match the current need often - at least in _my_ experience. 😉 When switching between searches, it would be useful to restore these options, too.

I am not thrilled. The options need to be added before the search string to not override options. There could be conflicts like -h vs -H too.
(not also that there is no --no-ignore-case option documented (but it exists) why overriding "case sensitive" is hard.)
So this would be annoying if you use long 
(I was fearing that the -e was to be removed which I would have liked less as other options would be hard to keep).

gerhardol commented on GitCommands/Settings/AppSettings.cs:
```diff
@@ -1354,6 +1360,12 @@ public static int DiffVerticalRulerPosition
             set => SetInt("diffverticalrulerposition", value);
         }
 
+        public static ISetting<string> GitGrepUserArguments { get; } = Setting.Create(AppearanceSettingsPath, nameof(GitGrepUserArguments), "");
+
+        public static ISetting<bool> GitGrepIgnoreCase { get; } = Setting.Create(AppearanceSettingsPath, nameof(GitGrepIgnoreCase), false);
+
+        public static ISetting<bool> GitGrepMatchWholeWord { get; } = Setting.Create(AppearanceSettingsPath, nameof(GitGrepMatchWholeWord), false);
```
can you clarify?

mstv replied:
git-grep tailoring should not be put into the category "appearance settings". Rather create a new category "dialog settings". It's time for proposing a better name.

gerhardol replied:
This was not changed, I have not got a better idea

mstv commented:
> The options need to be added before the search string to not override options. There could be conflicts like -h vs -H too.

It was just my 2ct. As it would add complexity, keep it as it is.

gerhardol commented on GitCommands/Git/GitModule.cs:
```diff
@@ -2297,6 +2297,81 @@ public string GetStatusText(bool untracked)
             });
         }
 
+        private ExecutionResult GetGrepFiles(ObjectId id, string grepString, CancellationToken cancellationToken = default)
```
There are qute some use of "id" too

gerhardol commented:
> This should be indicated by making the Search button the default button. This will also avoid the annoying Ding sound.
Then the label "Search regex:" can use the mnemonic 'S'; plus &Options:.

ok

> This popup should inherit from Igor's new default dialog -
most likely with Show search box in the footer.

ok

> The popup title should be "Search files in commit", too.
Or perhaps rather "Search in files at commit [using a regular expression...]" everywhere.

Using `Search files in commit`

> The history contains "-e". It should also remember the other options including case and whole word.
> It was just my 2ct. As it would add complexity, keep it as it is.

Yes, for now

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.Designer.cs:
```diff
@@ -434,6 +436,14 @@ private void InitializeComponent()
+            showSearchCommitToolStripMenuItem.Image = Properties.Images.ViewFile;
+            showSearchCommitToolStripMenuItem.Name = "showSearchCommitToolStripMenuItem";
+            showSearchCommitToolStripMenuItem.Size = new Size(262, 22);
+            showSearchCommitToolStripMenuItem.Text = "S&earch commit...";
```
I preferred the short but changed as suggested in the summary

gerhardol commented on UnitTests/GitCommands.Tests/Settings/AppSettingsTests.cs:
```diff
@@ -290,7 +290,12 @@ static IEnumerable<(PropertyInfo property, object defaultValue, bool isNullable,
                 yield return (properties[nameof(AppSettings.DiffVerticalRulerPosition)], 0, false, false);
+                yield return (properties[nameof(AppSettings.GitGrepUserArguments)], "", false, false);
+                yield return (properties[nameof(AppSettings.GitGrepIgnoreCase)], false, false, false);
+                yield return (properties[nameof(AppSettings.GitGrepMatchWholeWord)], false, false, false);
```
const bool isISetting = true;
const bool isNoISetting = false;
const bool isNotNullable = false;

gerhardol commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -1554,11 +1564,12 @@ public void SearchCommit_Click(string text)
-            search.Show();
+            _searchCommitForm.SearchFor = !string.IsNullOrEmpty(text) ? text : SearchComboBox.Text;
+            _searchCommitForm.SearchItems = SearchComboBox.Items;
+            _searchCommitForm.Location = new Point(TopLevelControl.Location.X + 100, TopLevelControl.Location.Y + 100);
```
The position is the same as for Find (should be offset a few pixel so they do not overlap)

gerhardol commented on GitUI/Editor/SearchCommitForm.cs:
```diff
@@ -22,7 +23,16 @@ public string SearchFor
-    public ComboBox.ObjectCollection SearchItems { private get; set; }
+    public ComboBox.ObjectCollection SearchItems
+    {
+        set
+        {
+            foreach (object item in value)
+            {
+                txtSearchFor.Items.Add(item);
+            }
+        }
```
Why a method for standard add here?

mstv replied:
> Why a method for standard add here?

Because "add" / "append" are not "set", which means "replace" somehow.

gerhardol commented on GitUI/Editor/SearchCommitForm.cs:
```diff
@@ -2,11 +2,12 @@
 
-    public SearchCommitForm()
+    public SearchCommitForm(GitUICommands commands)
+            : base(commands, enablePositionRestore: false)
```
Handled the same as find
I believe this is better, you want this related to the main window, not look for it when you have moved the main window.

mstv commented on GitCommands/Settings/AppSettings.cs:
```diff
@@ -1342,11 +1343,7 @@ public static bool ShowDiffForAllParents
-            get => GetBool("showsearchcommit", false);
-            set => SetBool("showsearchcommit", value);
-        }
+        public static ISetting<bool> ShowSearchCommit { get; } = Setting.Create(DialogSettingsPath, nameof(ShowSearchCommit), false);
```
This one is different.
```suggestion
        public static ISetting<bool> ShowSearchCommit { get; } = Setting.Create(AppearanceSettingsPath, nameof(ShowSearchCommit), false);
```

gerhardol commented:
> Because "add" / "append" are not "set", which means "replace" somehow.

This is a set operation. I changed the property to a method

mstv reviewed (commented):
Final nits on the UX:

![image](https://github.com/gitextensions/gitextensions/assets/36601201/2ecd4a68-70f5-4b9b-9286-c1057fff7ac4)

The search field should be focused on show.

The texts of "Show search box" and of the "Search" button should be aligned.

I guess it would be a clearer UX if the history entry becomes `-e "searchstring"` at once as in the search box.

![image](https://github.com/gitextensions/gitextensions/assets/36601201/205b079d-f949-4400-add4-85ceb18d0759)

... but only the history and not the input.
Otherwise, the input manipulation needs to update the cursor position.

Slowly typing 1 and 2 results in ![image](https://github.com/gitextensions/gitextensions/assets/36601201/2fe0a43b-1743-422b-92e4-16ec0cd06f1d)

gerhardol commented:
> The search field should be focused on show.

done

> The texts of "Show search box" and of the "Search" button should be aligned.

Cannot be set in a GitExtensionsDialog

> I guess it would be a clearer UX if the history entry becomes `-e "searchstring"` at once as in the search box.

The processing isin FileStatusList now and I would prefer to not expose this in the popup. But will consider changing

> ... but only the history and not the input. Otherwise, the input manipulation needs to update the cursor position.
> 
> Slowly typing 1 and 2 results in 

Not yet, should behandled

gerhardol commented:
> > I guess it would be a clearer UX if the history entry becomes `-e "searchstring"` at once as in the search box.
> 
> The processing is in FileStatusList now and I would prefer to not expose this in the popup. But will consider changing
> 
> Edit: can maybe overwrite from file status when updating

Aligned the dropdowns, the FileStatusList code updates both control.
Also change to keep "what is typed" in the dropdown. It is easier to edit in most situations.

> > ... but only the history and not the input. Otherwise, the input manipulation needs to update the cursor position.
> > Slowly typing 1 and 2 results in
> 
> Not yet, should be handled

When a search string was in the list and typed again, the item is moved first in the list. The text was updated too (due to added -e) which moved the cursor first that caused the problem.

gerhardol commented:
offset a few pixels compared to FindAndReplaceForm

![image](https://github.com/gitextensions/gitextensions/assets/6248932/86244313-bd9a-4d75-91c2-0d77637d2369)

mstv commented on GitUI/Editor/SearchCommitForm.cs:
```diff
@@ -0,0 +1,122 @@
+
+namespace GitUI;
+
+public partial class SearchCommitForm : GitExtensionsDialog
```
❔ The vast majority of the form names start with `Form`. 
```suggestion
public partial class FormSearchCommit : GitExtensionsDialog
```

gerhardol replied:
Almost no form in Editor has Form prefix. This was named in analogy with `FindAndReplaceForm`
Will rename though.

mstv reviewed (commented):
works better

Choosing an entry from the dropdown
- clears the search string of the FileStatusList
- in the Search form affects the FileStatusList at once - intentionally?
- should set the cursor behind the search string
- using cursor keys is not possible (only the first two entries can be reached)

Activating the Search button should keep the cursor position.

gerhardol commented:
Combobox handling is complex - the toolbar filter comboboxes are horrible.

> Choosing an entry from the dropdown
> 
>     * clears the search string of the FileStatusList

Added back the text assignment in dropdown update, but moving the cursor to the end.

>     * in the Search form affects the FileStatusList at once - intentionally?

It was intentional, but I changed it to avoid the strange workarounds in FormSearchCommit

>     * should set the cursor behind the search string

This occurred when the dropdown was updated and in the list but not the latest.

>     * using cursor keys is not possible (only the first two entries can be reached)

It is either keeping the list as a last-recently used sorted (this toggles the order) or just having a fifo queue.

> Activating the Search button should keep the cursor position.

In the popup, occurred when the list was resorted.

The list is not saved  in settings and I set it to 30 items, it is maybe not a problem to have it as fifo. Maybe even skip a maximal size, then most special handling can be removed?

mstv commented:
> Combobox handling is complex

Yes, it is.

It works much better now. I would not switch to FIFO mode and throw away the working code.

Remaining point: When changing the search string using the FileStatusList, only the dropdown of the SearchForm combobox is updated but not its text.

mstv commented:
👍 35bdaece2aa955f03c10a5327ffe7a90afadaa09 works well.

Just focus the search form on context menu and hotkey yet, please.

gerhardol commented:
> Just focus the search form on context menu and hotkey yet, please.

Do you mean when the popup is already open?
When opening the first time, the searchbox is focused
(when using menu/hotkey the second time the content is updated from the searchbox since the last update, also when not visible...)

mstv commented:
I mean 563c1cf2175ec7975b1ae57f82f52d4071df8fb5

gerhardol commented:
> I mean [563c1cf](https://github.com/gitextensions/gitextensions/commit/563c1cf2175ec7975b1ae57f82f52d4071df8fb5)

Thanks (you could have pushed to my branch directly, if I had not liked it I would revert...)

Setting Location was intentional, to avoid that the popup got lost.
But less risk the popup is lost as the popup is visible once at least.

Also a minor tweak.

I am sure there will be more tweaks to the functionality when there are more users

gerhardol commented:
> Setting Location was intentional, to avoid that the popup got lost. But less risk the popup is lost as the popup is visible once at least.

Would like to change this back so when using the hotkey/menu to invoke git-grep popup, show it at the default location.
If you see it, you likely not try to show it again. (Only reason would be that the current selection is added to the search box)

mstv commented:
> If you see it, you likely not try to show it again.

My usecase: Fingers at keyboard and wanting to focus the search form using `Shift+Ctrl+F`. 1. Focusing must work (had not before my commit). 2. Form should not jump away (not too important because I don't know how often I will move the form nearer to the FileStatusList).

gerhardol commented:
> > If you see it, you likely not try to show it again.
> 
> My usecase: Fingers at keyboard and wanting to focus the search form using `Shift+Ctrl+F`. 1. Focusing must work (had not before my commit). 2. Form should not jump away (not too important because I don't know how often I will move the form nearer to the FileStatusList).

Not how I use it, but I rather accept this and move on.
(Personally, I am using the searchbox. If find-in-diff was quicker I would like to have similar search there.)

RussKie commented:
@gerhardol do we have any docs on how to use this feature? It took me awhile to find that I have to use <kbd>CTRL + SHIFT + F</kbd> to open the new dialog, and I'm fairly certain our users won't find that easily.

RussKie commented:
Is this correct placeholder text? Should it be "...using git-grep"?
![image](https://github.com/user-attachments/assets/d63a21d6-b6bd-4a7c-ab60-9fd2b4ffa384)

RussKie commented:
> Mentioning the Git concept could be good, but "Search files in commit with gith git-grep using a regular expression" is quite long

This is a good description for the docs. 😉
